source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+    uint32_t operand_id = inst->GetSingleWordInOperand(i);
+    analysis::Const* operand_const = FindRecordedConst(operand_id);
+    if (!operand_const) return nullptr;
+    const analysis::Type* operand_type = operand_const->type();
+    assert(operand_type->AsVector() &&
+           "The first two operand of VectorShuffle must be of vector type");
+    if (analysis::VectorConst* vec_const = operand_const->AsVectorConst()) {
+      // case 1: current operand is a non-null vector constant.
+      for (uint32_t j = 0; j < vec_const->NumComponents(); j++) {
+        concatenated_components.push_back(vec_const->GetComponent(j));
+      }
+    } else if (operand_const->AsNullConst()) {
+      // case 2: current operand is a null vector constant. Create a temporary
+      // null scalar constant as the component.
+      const analysis::Type* component_type =
+          operand_type->AsVector()->element_type();
+      auto component_const = CreateConst(component_type, {});
+      // Append the null scalar consts to the concatenated components
+      // vector.
+      concatenated_components.insert(concatenated_components.end(),
+                                     operand_type->AsVector()->element_count(),
+                                     component_const.get());
+      // Insert an instruction for the component constant.
+      // TODO(qining): Here we may insert an instruction while return a null
+      // pointer. In that case, the process() will return false and assumes the
+      // module is not modified, but actually it is modified. Need to fix it.
+      BuildInstructionAndAddToModule(std::move(component_const), pos);
+    } else {
+      // no other valid cases
+      return nullptr;
+    }
+  }
+  // Create the new vector constant with the selected components.
+
+  std::vector<const analysis::Const*> selected_components;
+  for (uint32_t i = 3; i < inst->NumInOperands(); i++) {
+    assert(inst->GetInOperand(i).type == SPV_OPERAND_TYPE_LITERAL_INTEGER &&
+           "The literal operand must of type SPV_OPERAND_TYPE_LITERAL_INTEGER");
+    uint32_t literal = inst->GetSingleWordInOperand(i);
+    assert(literal < concatenated_components.size() &&
+           "Literal index out of bound of the concatenated vector");
+    selected_components.push_back(concatenated_components[literal]);
+  }
+  auto new_vec_const =
+      MakeUnique<analysis::VectorConst>(result_vec_type, selected_components);
+  return BuildInstructionAndAddToModule(std::move(new_vec_const), pos);
+}
+
+namespace {
+// A helper function that traverses the id operands of the given
+// OpSpecConstantOp instruction and run a given checker function |f| for each
+// id. Returns true if the results of |f| with all operand ids return true,
+// otherwise return false.
+// The function |f| is used as a checker. It takes an operand id and returns
+// bool to indicate the checking result.
+bool AllOperandsSatisfy(const ir::Instruction& inst,
```

Comment:
If an instruction had operand iterators, then this would be better as an application of
std::all_of

Please add a TODO to replace this.  Or just replace this now.